tests/stack/test_api.py

```diff
@@ -529,3 +572,83 @@ def test_download_resource(self, stack_api):
+            resource_properties = resource[api.RESOURCE_PROPERTIES]
+            curr_source_path = resource_properties.get(api.DBFS_RESOURCE_SOURCE_PATH, '')
+            resource_properties.update(
+                {api.DBFS_RESOURCE_SOURCE_PATH: os.path.join(tmpdir.strpath, curr_source_path)})
+            if resource_service == api.WORKSPACE_SERVICE:
+                if workspace_api.NOTEBOOK == \
+                        resource_properties[api.WORKSPACE_RESOURCE_OBJECT_TYPE]:
+                    os.makedirs(os.path.dirname(resource_properties[
+                                                api.WORKSPACE_RESOURCE_SOURCE_PATH]))
+                    with open(resource_properties[api.WORKSPACE_RESOURCE_SOURCE_PATH], 'w') as f:
+                        f.write("print('test')\n")
+                if resource_properties[api.WORKSPACE_RESOURCE_OBJECT_TYPE] == \
+                        workspace_api.DIRECTORY:
+                    os.makedirs(resource_properties[api.WORKSPACE_RESOURCE_SOURCE_PATH])
+            elif resource_service == api.DBFS_SERVICE:
+                if resource_properties[api.DBFS_RESOURCE_IS_DIR]:
+                    os.makedirs(resource_properties[api.DBFS_RESOURCE_SOURCE_PATH])
+                else:
+                    with open(resource_properties[api.DBFS_RESOURCE_SOURCE_PATH], 'w') as f:
+                        f.write("print('test')\n")
+
+        # Run deploy command on stack and validate stack status. If there are no exceptions
+        # related to validation errors for the newly generated stack status, then test
+        # will pass.
+        new_stack_status_1 = stack_api.deploy_config(test_stack)
+        # Test some extra correctness criteria for the generated stack status.
+        assert new_stack_status_1.get(api.STACK_RESOURCES) == test_stack.get(api.STACK_RESOURCES)
+        for deployed_resource in new_stack_status_1.get(api.STACK_DEPLOYED):
+            assert deployed_resource.get(api.RESOURCE_DEPLOY_OUTPUT) == test_deploy_output
+
+        # Test that when passing in a  status that the new status generated will still be valid.
+        new_stack_status_2 = stack_api.deploy_config(test_stack, stack_status=TEST_STATUS)
+        # Test some extra correctness criteria for the generated stack status.
+        assert new_stack_status_2.get(api.STACK_RESOURCES) == test_stack.get(api.STACK_RESOURCES)
+        for deployed_resource in new_stack_status_2.get(api.STACK_DEPLOYED):
+            assert deployed_resource.get(api.RESOURCE_DEPLOY_OUTPUT) == test_deploy_output
+
+    def test_download(self, stack_api):
+        """
+            stack_api.download_from_config should call the correct workspace_client endpoints
+            when the a valid input is given.
+        """
+        stack_api.workspace_client.export_workspace = mock.MagicMock()
+        stack_api.workspace_client.export_workspace_dir = mock.MagicMock()
+
+        stack_api.download_from_config(TEST_STACK)
```

Comment:
It would be better to separate those into two tests, because it may be confusing and hard to debug if you count the total number of invocations of the REST API methods.

Reply:
I'm going to actually just make this test count the number of times that `_download_workspace` is called. I think that will be easier to debug.